### ISSUE TITLE
Detailed Guidance for setup of CLAIRCTL

### ISSUE THREAD

**Issue by obikay200** [jgsqware/clairctl#18]
I very much like the work you guys are doing on this,

Could you please provide some instruction on how to setup CLAIRCTL and the requirements for its running environment?

Thanks

**Comment by jgsqware** [comment]
Thank for the interest. The only thing you need to configure is the .clairctl.yml file as explained in the readme with the uri and port of you clair.

then, it will work out of box

**Comment by obikay200** [comment]
Thanks for the quick response, that working ok now, however when i try to analyse an image in the docker hub i get th following:

# ./clairctl analyze obikay200/loginapp:v2
client quit unexpectedly
FATA[0002] pushing image "obikay200/loginapp:v2": receiving http error: 400 
# ping clair    
PING clair (172.19.0.3): 56 data bytes
64 bytes from 172.19.0.3: icmp_seq=0 ttl=64 time=0.048 ms
64 bytes from 172.19.0.3: icmp_seq=1 ttl=64 time=0.048 ms
64 bytes from 172.19.0.3: icmp_seq=2 ttl=64 time=0.068 ms
^C--- clair ping statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max/stddev = 0.048/0.055/0.068/0.000 ms
# ./clairctl health
Clair: ✔

I can Pull ok:
# ./clairctl pull obikay200/loginapp:v2

Image: docker.io/obikay200/loginapp:v2
 5 layers found
  ➜ sha256:b7f33cc0b48ea4fb2f0745def58c25483a5f6b7aed5b41ce8f1cb6e17f5723cf
  ➜ sha256:9e0a04370e7ea9bc8b898558174588d278816dc50f008a09163e94a897f84ad2
  ➜ sha256:be4e8f4f78343d5b9830b73062afcf83d4547658c12cc96bc22c7aefa15f0c06
  ➜ sha256:ae289f37c57d122ea8caf825cac469bfd675895cadc0c6ddc5657525d8d06e96
  ➜ sha256:7018c030de099695c2d3fe47f855749a9917af4767d529cb0ee928ef99e2e163
 
But i get the same error on push:

# ./clairctl push obikay200/loginapp:v2
client quit unexpectedly
FATA[0002] pushing image "obikay200/loginapp:v2": receiving http error: 400 

# ./clairctl analyze obikay200/loginapp:v2
client quit unexpectedly
2017-02-28 12:31:36.165243 C | cmd: pushing image "obikay200/loginapp:v2": receiving http error: 400

**Comment by jgsqware** [comment]
Do you have some log in clair?

http 400 means Bad Request

**Comment by obikay200** [comment]
2017-02-28 11:26:18.557721 I | api: starting main API on port 6060.
2017-02-28 11:26:18.557780 I | updater: updater service started. lock identifier: e004a897-a1c6-4fbe-9f49-a1e3b191497a
2017-02-28 11:26:18.569301 I | updater: updating vulnerabilities
2017-02-28 11:26:18.569350 I | updater: fetching vulnerability updates
2017-02-28 11:26:18.569393 I | updater/fetchers/ubuntu: fetching Ubuntu vulnerabilities
2017-02-28 11:26:18.569696 I | updater/fetchers/debian: fetching Debian vulnerabilities
2017-02-28 11:26:18.570142 I | updater/fetchers/rhel: fetching Red Hat vulnerabilities
2017-02-28 11:26:46.082123 I | main: Received interruption, gracefully stopping ...
2017-02-28 11:26:46.082347 I | api: health API stopped
2017-02-28 11:26:46.082390 I | api: main API stopped
2017-02-28 11:26:46.084865 I | updater: updater service stopped
2017-02-28 11:26:46.591688 I | pgsql: running database migrations
goose: no migrations to run. current version: 20151222113213
2017-02-28 11:26:46.607187 I | pgsql: database migration ran successfully
2017-02-28 11:26:46.607572 I | notifier: notifier service is disabled
2017-02-28 11:26:46.607603 I | api: starting main API on port 6060.
2017-02-28 11:26:46.607869 I | api: starting health API on port 6061.
2017-02-28 11:26:46.607976 I | updater: updater service started. lock identifier: c45fc4b1-6e76-4712-bfbc-4ea90570ab85
2017-02-28 11:26:46.617474 I | updater: updating vulnerabilities
2017-02-28 11:26:46.617531 I | updater: fetching vulnerability updates
2017-02-28 11:26:46.617582 I | updater/fetchers/rhel: fetching Red Hat vulnerabilities
2017-02-28 11:26:46.617662 I | updater/fetchers/ubuntu: fetching Ubuntu vulnerabilities
2017-02-28 11:26:46.618281 I | updater/fetchers/debian: fetching Debian vulnerabilities
2017-02-28 11:27:19.002333 I | main: Received interruption, gracefully stopping ...
2017-02-28 11:27:19.002568 I | api: main API stopped
2017-02-28 11:27:19.002604 I | api: health API stopped
2017-02-28 11:27:19.005199 I | updater: updater service stopped
2017-02-28 11:27:19.541304 I | pgsql: running database migrations
goose: no migrations to run. current version: 20151222113213
2017-02-28 11:27:19.557952 I | pgsql: database migration ran successfully
2017-02-28 11:27:19.558357 I | notifier: notifier service is disabled
2017-02-28 11:27:19.558396 I | api: starting health API on port 6061.
2017-02-28 11:27:19.558522 I | api: starting main API on port 6060.
2017-02-28 11:27:19.558734 I | updater: updater service started. lock identifier: 6cc3cc3e-0782-4a03-ac1f-bdb782f7a028
2017-02-28 11:27:19.569293 I | updater: updating vulnerabilities
2017-02-28 11:27:19.569350 I | updater: fetching vulnerability updates
2017-02-28 11:27:19.569404 I | updater/fetchers/ubuntu: fetching Ubuntu vulnerabilities
2017-02-28 11:27:19.569996 I | updater/fetchers/debian: fetching Debian vulnerabilities
2017-02-28 11:27:19.570767 I | updater/fetchers/rhel: fetching Red Hat vulnerabilities
2017-02-28 11:40:07.933766 I | updater: adding metadata to vulnerabilities
2017-02-28 11:47:02.988656 W | detectors: could not download layer: got status code 401, expected 2XX
2017-02-28 11:47:02.988726 E | worker: layer sha256:b7f33cc0b48ea4fb2f0745def58c25483a5f6b7aed5b41ce8f1cb6e17f5723cf: failed to extract data from http://172.19.0.3:37263/v2/obikay200/loginapp/blobs/sha256:b7f33cc0b48ea4fb2f0745def58c25483a5f6b7aed5b41ce8f1cb6e17f5723cf: could not find layer
2017-02-28 11:47:02.989792 I | api: [::1]:40174 "POST /v1/layers" 400 (1.116578132s)
2017-02-28 11:50:51.010896 I | api: [::1]:53426 "GET /health" 200 (44.78µs)
2017-02-28 11:56:41.276109 W | detectors: could not download layer: got status code 401, expected 2XX
2017-02-28 11:56:41.276168 E | worker: layer sha256:b7f33cc0b48ea4fb2f0745def58c25483a5f6b7aed5b41ce8f1cb6e17f5723cf: failed to extract data from http://172.19.0.3:33149/v2/obikay200/loginapp/blobs/sha256:b7f33cc0b48ea4fb2f0745def58c25483a5f6b7aed5b41ce8f1cb6e17f5723cf: could not find layer
2017-02-28 11:56:41.277175 I | api: [::1]:40254 "POST /v1/layers" 400 (1.061883416s)
2017-02-28 11:59:16.332167 I | api: [::1]:40286 "GET /v1/layers/sha256:7018c030de099695c2d3fe47f855749a9917af4767d529cb0ee928ef99e2e163?vulnerabilities" 404 (1.131528ms)
2017-02-28 11:59:16.340001 I | api: [::1]:40286 "GET /v1/layers/sha256:ae289f37c57d122ea8caf825cac469bfd675895cadc0c6ddc5657525d8d06e96?vulnerabilities" 404 (7.207348ms)
2017-02-28 11:59:16.342212 I | api: [::1]:40286 "GET /v1/layers/sha256:be4e8f4f78343d5b9830b73062afcf83d4547658c12cc96bc22c7aefa15f0c06?vulnerabilities" 404 (1.689406ms)
2017-02-28 11:59:16.344380 I | api: [::1]:40286 "GET /v1/layers/sha256:9e0a04370e7ea9bc8b898558174588d278816dc50f008a09163e94a897f84ad2?vulnerabilities" 404 (1.756878ms)
2017-02-28 11:59:16.346341 I | api: [::1]:40286 "GET /v1/layers/sha256:b7f33cc0b48ea4fb2f0745def58c25483a5f6b7aed5b41ce8f1cb6e17f5723cf?vulnerabilities" 404 (1.616615ms)
2017-02-28 12:10:04.454630 I | updater: update finished
2017-02-28 12:29:02.224658 I | api: [::1]:53776 "GET /health" 200 (32.621µs)
2017-02-28 12:31:36.164037 W | detectors: could not download layer: got status code 401, expected 2XX
2017-02-28 12:31:36.164125 E | worker: layer sha256:b7f33cc0b48ea4fb2f0745def58c25483a5f6b7aed5b41ce8f1cb6e17f5723cf: failed to extract data from http://172.19.0.3:42576/v2/obikay200/loginapp/blobs/sha256:b7f33cc0b48ea4fb2f0745def58c25483a5f6b7aed5b41ce8f1cb6e17f5723cf: could not find layer
2017-02-28 12:31:36.165040 I | api: [::1]:40580 "POST /v1/layers" 400 (1.057407541s)
2017-02-28 12:46:11.119026 W | detectors: could not download layer: got status code 401, expected 2XX
2017-02-28 12:46:11.119085 E | worker: layer sha256:b7f33cc0b48ea4fb2f0745def58c25483a5f6b7aed5b41ce8f1cb6e17f5723cf: failed to extract data from http://172.19.0.3:41277/v2/obikay200/loginapp/blobs/sha256:b7f33cc0b48ea4fb2f0745def58c25483a5f6b7aed5b41ce8f1cb6e17f5723cf: could not find layer
2017-02-28 12:46:11.119978 I | api: [::1]:40680 "POST /v1/layers" 400 (1.027550494s)

**Comment by jgsqware** [comment]
is this ip resolvable from your clair instance: `http://172.19.0.3`

clairctl pplay as reverse proxy for authentication

**Comment by obikay200** [comment]
yeah im using the container image on the compose file from clairs github page

**Comment by obikay200** [comment]
clairctl is in the same container so they should speak ok

**Comment by jgsqware** [comment]
Ok I haven't test it from the same container, I will try that

**Comment by obikay200** [comment]
the  .clairctl.yml file you mention where does that sit?

and the content is something like

clair:
  port: 6060
  healthPort: 6061
  uri: http://clair
  report:
    path: ./reports
    format: html
docker:
  insecure-registries: 
    - "my-own-registry:5000"

**Comment by obikay200** [comment]
As in whats it default location?

**Comment by jgsqware** [comment]
you mean the default location of `clairctl.yml`? beside the clairctl binary or in your home folder

**Comment by jgsqware** [comment]
what is the result of `clairctl --log-level debug push obikay200/loginapp:v2`

**Comment by obikay200** [comment]
yes, could that sit in the same file or is it looking for it e.g. /config/clairctl.yml

Output

# ./clairctl --log-level debug push obikay200/loginapp:v2
2017-02-28 13:10:17.664048 D | config: Using config file: /config/clairctl.yml
2017-02-28 13:10:17.664296 I | config: retrieving interface for local IP
2017-02-28 13:10:17.664308 D | config: no interface provided, looking for docker0
2017-02-28 13:10:17.664866 D | config: docker0 not found, looking for first connected broadcast interface
2017-02-28 13:10:17.665445 D | server: Update local server port from "0" to "40664"
2017-02-28 13:10:17.665463 I | server: Starting Server on 172.19.0.3:40664
2017-02-28 13:10:17.670100 D | dockerdist: Downloading manifest for obikay200/loginapp:v2
2017-02-28 13:10:17.670307 D | dockerdist: Retrieving repository client
2017-02-28 13:10:17.685154 D | dockerdist: endpoint.TLSConfig.InsecureSkipVerify: true
2017-02-28 13:10:19.182663 D | dockerdist: manifest type: *schema2.DeserializedManifest
2017-02-28 13:10:19.182689 D | dockerdist: retrieved schema2 manifest, no verification
2017-02-28 13:10:19.182707 I | config: retrieving interface for local IP
2017-02-28 13:10:19.182711 D | config: no interface provided, looking for docker0
2017-02-28 13:10:19.182811 D | config: docker0 not found, looking for first connected broadcast interface
2017-02-28 13:10:19.182921 I | clair: Pushing Layer 1/5 [sha256:b7f33]
2017-02-28 13:10:19.182978 D | clair: Saving sha256:b7f33cc0b48ea4fb2f0745def58c25483a5f6b7aed5b41ce8f1cb6e17f5723cf[https://registry-1.docker.io]
2017-02-28 13:10:19.185355 D | server: request url: /v2/obikay200/loginapp/blobs/sha256:b7f33cc0b48ea4fb2f0745def58c25483a5f6b7aed5b41ce8f1cb6e17f5723cf
2017-02-28 13:10:19.185371 D | server: request for image: obikay200/loginapp:v2
2017-02-28 13:10:19.185407 D | server: host retreived: https://registry-1.docker.io
2017-02-28 13:10:19.185424 D | server: auth.insecureSkipVerify: true
2017-02-28 13:10:19.185433 D | server: request.URL.String(): https://registry-1.docker.io/v2/obikay200/loginapp/blobs/sha256:b7f33cc0b48ea4fb2f0745def58c25483a5f6b7aed5b41ce8f1cb6e17f5723cf
2017-02-28 13:10:19.533361 I | server: pull from clair is unauthorized
2017-02-28 13:10:20.243910 I | clair: adding layer 1/5 [sha256:b7f33]: receiving http error: 400
client quit unexpectedly
2017-02-28 13:10:20.243939 C | cmd: pushing image "obikay200/loginapp:v2": receiving http error: 400
#

**Comment by jgsqware** [comment]
are you in your clair container?

**Comment by obikay200** [comment]
yeah

**Comment by jgsqware** [comment]
Ok i've reproduced it.

If you want to use it now, before I fixed it, you can run clairctl from outside clair container, and refer in your clairctl.yml file your clair uri 

```
uri:
```

**Comment by obikay200** [comment]
so run from outside the container, on the host machine, then enter the ip of the container in the uri yeah

**Comment by jgsqware** [comment]
running from outside it's working?

**Comment by obikay200** [comment]
heres the output:

# ./clairctl --log-level debug push obikay200/loginapp:v2
2017-02-28 13:35:17.844456 D | config: Using config file: /root/clair_config/clairctl.yml
2017-02-28 13:35:17.844620 I | config: retrieving interface for local IP
2017-02-28 13:35:17.844627 D | config: no interface provided, looking for docker0
2017-02-28 13:35:17.846780 D | server: Update local server port from "0" to "35453"
2017-02-28 13:35:17.846846 I | server: Starting Server on 172.17.0.1:35453
2017-02-28 13:35:17.851193 D | dockerdist: Downloading manifest for obikay200/loginapp:v2
2017-02-28 13:35:17.851412 D | dockerdist: Retrieving repository client
2017-02-28 13:35:19.047035 D | dockerdist: endpoint.TLSConfig.InsecureSkipVerify: true
2017-02-28 13:35:20.625155 D | dockerdist: manifest type: *schema2.DeserializedManifest
2017-02-28 13:35:20.625201 D | dockerdist: retrieved schema2 manifest, no verification
2017-02-28 13:35:20.625223 I | config: retrieving interface for local IP
2017-02-28 13:35:20.625228 D | config: no interface provided, looking for docker0
2017-02-28 13:35:20.625549 I | clair: Pushing Layer 1/5 [sha256:b7f33]
2017-02-28 13:35:20.625619 D | clair: Saving sha256:b7f33cc0b48ea4fb2f0745def58c25483a5f6b7aed5b41ce8f1cb6e17f5723cf[https://registry-1.docker.io]
2017-02-28 13:35:20.629997 D | server: request url: /v2/obikay200/loginapp/blobs/sha256:b7f33cc0b48ea4fb2f0745def58c25483a5f6b7aed5b41ce8f1cb6e17f5723cf
2017-02-28 13:35:20.630010 D | server: request for image: obikay200/loginapp:v2
2017-02-28 13:35:20.630054 D | server: host retreived: https://registry-1.docker.io
2017-02-28 13:35:20.630071 D | server: auth.insecureSkipVerify: true
2017-02-28 13:35:20.630078 D | server: request.URL.String(): https://registry-1.docker.io/v2/obikay200/loginapp/blobs/sha256:b7f33cc0b48ea4fb2f0745def58c25483a5f6b7aed5b41ce8f1cb6e17f5723cf
2017-02-28 13:35:20.971497 I | server: pull from clair is unauthorized
2017-02-28 13:35:21.916933 I | clair: Pushing Layer 2/5 [sha256:9e0a0]
2017-02-28 13:35:21.917019 D | clair: Saving sha256:9e0a04370e7ea9bc8b898558174588d278816dc50f008a09163e94a897f84ad2[https://registry-1.docker.io]
2017-02-28 13:35:21.922870 D | server: request url: /v2/obikay200/loginapp/blobs/sha256:9e0a04370e7ea9bc8b898558174588d278816dc50f008a09163e94a897f84ad2
2017-02-28 13:35:21.922892 D | server: request for image: obikay200/loginapp:v2
2017-02-28 13:35:21.922931 D | server: host retreived: https://registry-1.docker.io
2017-02-28 13:35:21.922953 D | server: auth.insecureSkipVerify: true
2017-02-28 13:35:21.922961 D | server: request.URL.String(): https://registry-1.docker.io/v2/obikay200/loginapp/blobs/sha256:9e0a04370e7ea9bc8b898558174588d278816dc50f008a09163e94a897f84ad2
2017-02-28 13:35:22.269256 I | server: pull from clair is unauthorized
2017-02-28 13:35:24.797477 I | clair: Pushing Layer 3/5 [sha256:be4e8]
2017-02-28 13:35:24.797560 D | clair: Saving sha256:be4e8f4f78343d5b9830b73062afcf83d4547658c12cc96bc22c7aefa15f0c06[https://registry-1.docker.io]
2017-02-28 13:35:24.803106 D | server: request url: /v2/obikay200/loginapp/blobs/sha256:be4e8f4f78343d5b9830b73062afcf83d4547658c12cc96bc22c7aefa15f0c06
2017-02-28 13:35:24.803126 D | server: request for image: obikay200/loginapp:v2
2017-02-28 13:35:24.803191 D | server: host retreived: https://registry-1.docker.io
2017-02-28 13:35:24.803211 D | server: auth.insecureSkipVerify: true
2017-02-28 13:35:24.803219 D | server: request.URL.String(): https://registry-1.docker.io/v2/obikay200/loginapp/blobs/sha256:be4e8f4f78343d5b9830b73062afcf83d4547658c12cc96bc22c7aefa15f0c06
2017-02-28 13:35:25.148416 I | server: pull from clair is unauthorized
2017-02-28 13:35:26.129979 I | clair: Pushing Layer 4/5 [sha256:ae289]
2017-02-28 13:35:26.130088 D | clair: Saving sha256:ae289f37c57d122ea8caf825cac469bfd675895cadc0c6ddc5657525d8d06e96[https://registry-1.docker.io]
2017-02-28 13:35:26.134356 D | server: request url: /v2/obikay200/loginapp/blobs/sha256:ae289f37c57d122ea8caf825cac469bfd675895cadc0c6ddc5657525d8d06e96
2017-02-28 13:35:26.134380 D | server: request for image: obikay200/loginapp:v2
2017-02-28 13:35:26.134428 D | server: host retreived: https://registry-1.docker.io
2017-02-28 13:35:26.134447 D | server: auth.insecureSkipVerify: true
2017-02-28 13:35:26.134456 D | server: request.URL.String(): https://registry-1.docker.io/v2/obikay200/loginapp/blobs/sha256:ae289f37c57d122ea8caf825cac469bfd675895cadc0c6ddc5657525d8d06e96
2017-02-28 13:35:26.477738 I | server: pull from clair is unauthorized
2017-02-28 13:35:27.940690 I | clair: Pushing Layer 5/5 [sha256:7018c]
2017-02-28 13:35:27.940821 D | clair: Saving sha256:7018c030de099695c2d3fe47f855749a9917af4767d529cb0ee928ef99e2e163[https://registry-1.docker.io]
2017-02-28 13:35:27.945220 D | server: request url: /v2/obikay200/loginapp/blobs/sha256:7018c030de099695c2d3fe47f855749a9917af4767d529cb0ee928ef99e2e163
2017-02-28 13:35:27.945239 D | server: request for image: obikay200/loginapp:v2
2017-02-28 13:35:27.945290 D | server: host retreived: https://registry-1.docker.io
2017-02-28 13:35:27.945309 D | server: auth.insecureSkipVerify: true
2017-02-28 13:35:27.945318 D | server: request.URL.String(): https://registry-1.docker.io/v2/obikay200/loginapp/blobs/sha256:7018c030de099695c2d3fe47f855749a9917af4767d529cb0ee928ef99e2e163
2017-02-28 13:35:28.305818 I | server: pull from clair is unauthorized
obikay200/loginapp:v2 has been pushed to Clair

**Comment by obikay200** [comment]
yeah thats working i think, i have a report

**Comment by obikay200** [comment]
Thanks for the great work!

**Comment by jgsqware** [comment]
ok cool thanks! I will investigate why it's not working from the same container as Clair

**Comment by farhanible** [comment]
I'm having this same issue but running clairctl in a separate container. The logs messages on both ClairCTL and Clair are the same.

`docker run --network ubuntu_default --link clair:clair -v $PWD/clair_config:/config chussenot/clairctl ./clairctl --log-level Debug --config=/config/clairctl.yml push library/nginx`

Results in:

```
2017-05-31 21:16:41.630954 D | config: Using config file: /config/clairctl.yml
2017-05-31 21:16:41.631486 D | dockerdist: Downloading manifest for library/nginx
2017-05-31 21:16:41.631791 D | dockerdist: Retrieving repository client
2017-05-31 21:16:41.632080 D | dockerdist: endpoint.TLSConfig.InsecureSkipVerify: true
2017-05-31 21:16:41.789068 D | dockerdist: manifest type: *schema2.DeserializedManifest
2017-05-31 21:16:41.789453 D | dockerdist: retrieved schema2 manifest, no verification
2017-05-31 21:16:41.789626 I | config: retrieving interface for local IP
2017-05-31 21:16:41.789776 D | config: no interface provided, looking for docker0
2017-05-31 21:16:41.790017 D | config: docker0 not found, looking for first connected broadcast interface
2017-05-31 21:16:41.790339 I | clair: Pushing Layer 1/3 [sha256:ff3d5]
2017-05-31 21:16:41.790578 D | clair: Saving sha256:ff3d52d8f55fb0b74ea0a24134f75efeff780c4e1f407073def2ae9c9b900868[https://registry-1.docker.io/v2]
2017-05-31 21:16:41.807115 I | clair: adding layer 1/3 [sha256:ff3d5]: receiving http error: 400
client quit unexpectedly
2017-05-31 21:16:41.807436 C | cmd: pushing image "nginx:latest": receiving http error: 400

```

The Clair side of the logs are:

```
{"Event":"could not download layer: expected 2XX","Level":"warning","Location":"driver.go:135","Time":"2017-05-31 21:16:41.805966","status code":401}
{"Event":"failed to extract data from path","Level":"error","Location":"worker.go:122","Time":"2017-05-31 21:16:41.806049","error":"could not find layer","layer":"sha256:ff3d52d8f55fb0b74ea0a24134f75efeff780c4e1f407073def2ae9c9b900868","path":"https://registry-1.docker.io/v2/library/nginx/blobs/sha256:ff3d52d8f55fb0b74ea0a24134f75efeff780c4e1f407073def2ae9c9b900868"}
{"Event":"Handled HTTP request","Level":"info","Location":"router.go:57","Time":"2017-05-31 21:16:41.806776","elapsed time":14415658,"method":"POST","remote addr":"172.18.0.4:41198","request uri":"/v1/layers","status":"400"}

```

**Comment by farhanible** [comment]
Looks like Docker Hub requires authentication. Is it possible to specify authentication in Clairctl so that it fetches and sends layers to Clair? Or is it just enumerating layers and sending URIs to Clair, which means Clair would need credentials to fetch the layers.

`{"errors":[{"code":"UNAUTHORIZED","message":"authentication required","detail":[{"Type":"repository","Class":"","Name":"library/nginx","Action":"pull"}]}]}`

**Comment by jgsqware** [comment]
Yes it use the docker credentials, so if you logged into docker you will
logged with clairctl

If you don't have docker on your image you can mount your local
~/.docker/config or something like that

I will redo the login command in clairctl it self for standalone application

**Comment by farhanible** [comment]
> If you don't have docker on your image you can mount your local
> ~/.docker/config or something like that

To clairctl or clair?

**Comment by farhanible** [comment]
From the Clair [documentation](https://github.com/coreos/clair/blob/master/Documentation/api_v1.md#post-layers): 

> The Authorization field is an optional value whose contents will fill the Authorization HTTP Header when requesting the layer via HTTP.

So an authorization header can be injected into the /layers POST request. I believe that is so that Clair can insert that Authorization header into it's own calls to download the layer tarballs.

I can't find where this key value pair is being injected in clairctl. I assume that if clairctl is acting as a reverse proxy for clair calls, the layer path being passed to clair would look like
 
`http://clairctl/nginx/blobs/sha256:ff3d52d8f55fb0b74ea0a24134f75efeff780c4e1f407073def2ae9c9b900868`

 instead of 

`https://registry-1.docker.io/v2/library/nginx/blobs/sha256:ff3d52d8f55fb0b74ea0a24134f75efeff780c4e1f407073def2ae9c9b900868
`

Let me know what I'm doing wrong here. Thanks for your help.

**Comment by jgsqware** [comment]
You're right. But in the last version, the authentication header is set
before sending the request to clair. So if you logged in to docker you
should be logged

Could you show the output in debug level?

**Comment by riponbanik** [comment]
@jgsqware The issue is still present, I am running from outside, 

 clairctl --log-level debug analyze xxx/k8s-app:b1
2020-07-07 09:52:54.256996 D | config: Using config file: /home/ripon/workspace/private/kubernetes/security/clair/clairctl.yml
2020-07-07 09:52:54.257463 D | dockerdist: Downloading manifest for ripon/k8s-app:b1
2020-07-07 09:52:54.257640 D | dockerdist: Retrieving repository client
2020-07-07 09:52:57.072449 D | dockerdist: endpoint.TLSConfig.InsecureSkipVerify: true
2020-07-07 09:53:00.835138 D | dockerdist: manifest type: *schema2.DeserializedManifest
2020-07-07 09:53:00.835384 D | dockerdist: retrieved schema2 manifest, no verification
2020-07-07 09:53:00.835594 I | config: retrieving interface for local IP
2020-07-07 09:53:00.835638 D | config: no interface provided, looking for docker0
2020-07-07 09:53:00.837037 I | clair: Pushing Layer 1/6 [sha256:fc718]
2020-07-07 09:53:00.837375 D | clair: Saving sha256:fc7181108d403205fda45b28dbddfa1cf07e772fa41244e44f53a341b8b1893d[https://registry-1.docker.io/v2]
2020-07-07 09:53:00.837648 D | clair: auth.insecureSkipVerify: true
2020-07-07 09:53:00.837779 D | clair: request.URL.String(): https://registry-1.docker.io/v2/ripon/k8s-app/blobs/sha256:fc7181108d403205fda45b28dbddfa1cf07e772fa41244e44f53a341b8b1893d
2020-07-07 09:53:01.769218 I | clair: pull from clair is unauthorized
2020-07-07 09:53:12.718141 I | clair: adding layer 1/6 [sha256:fc718]: receiving http error: 400
client quit unexpectedly
2020-07-07 09:53:12.718241 C | cmd: pushing image "xxx/k8s-app:b1": receiving http error: 400
